### PR TITLE
[Package Signing] Integrate the ICertificateVerifier

### PR DESCRIPTION
**Pull request by loic-sharma**
Integrates the `ICertificateVerifier` to the Validate Certificate job. The `ICertificateVerifier`'s implementation depends on https://github.com/NuGet/NuGet.Jobs/pull/309.

### CODE REVIEW

File: src/Validation.PackageSigning.ValidateCertificate/CertificateValidationService.cs [hunk 1]
```diff
@@ -43,19 +43,15 @@ public class CertificateValidationService : ICertificateValidationService
 
         public Task<EndCertificateValidation> FindCertificateValidationAsync(CertificateValidationMessage message)
         {
+            // Fetch the validation, the end certificate that this validation is for, and all of the parent
+            // certificates that the end certificate depends on.
             return _context
                         .CertificateValidations
                         .Where(v => v.ValidationId == message.ValidationId && v.EndCertificateKey == message.CertificateKey)
-                        .Include(v => v.EndCertificate)
+                        .Include(v => v.EndCertificate.CertificateChainLinks.Select(l => l.EndCertificate))
```
Comment (dtivel) [inline src/Validation.PackageSigning.ValidateCertificate/CertificateValidationService.cs]: It seems like the 2nd lambda should be `l => l.ParentCertificate`.  (I'm guessing at the name of `ParentCertificate`.)

File: src/Validation.PackageSigning.ValidateCertificate/CertificateValidationMessageHandler.cs [hunk 2]
```diff
@@ -96,8 +102,27 @@ public sealed class CertificateValidationMessageHandler : IMessageHandler<Certif
             }
 
             // Download and verify the certificate.
```
Comment (dtivel) [inline src/Validation.PackageSigning.ValidateCertificate/CertificateValidationMessageHandler.cs]: Remove the comment.  It's at least outdated, but the subsequent statement is awesomely clear as to what it does by virtue of good naming.

File: src/Validation.PackageSigning.ValidateCertificate/SignatureDeciderFactory.cs [hunk 3]
```diff
@@ -56,14 +56,13 @@ public SignatureDecider MakeDeciderForInvalidatedCertificate(EndCertificate cert
...
-            if (result.StatusFlags == X509ChainStatusFlags.NoError ||
-                (result.StatusFlags & (X509ChainStatusFlags.OfflineRevocation | X509ChainStatusFlags.RevocationStatusUnknown)) != 0)
+            if (result.StatusFlags == X509ChainStatusFlags.NoError)
             {
                 throw new ArgumentException($"Invalid flags on invalid verification result: {result.StatusFlags}!", nameof(result));
             }
 
             // If a certificate used for the primary signature is revoked, all dependent signatures should be invalidated.
-            // Note that in this case the revoked certificate is a parent of the end certificate - NOT the end certificate.
+            // NOTE: It is assumed that the revoked certificate is a parent certificate, but this may not be strictly true.
```
Comment (dtivel) [inline src/Validation.PackageSigning.ValidateCertificate/SignatureDeciderFactory.cs]: Parent or ancestor?

File: src/Validation.PackageSigning.ValidateCertificate/CertificateValidationMessageHandler.cs [hunk 4]
```diff
@@ -155,5 +180,31 @@ private bool HasValidationCompleted(EndCertificateValidation validation, Certifi
 
             throw new InvalidOperationException($"Unknown {nameof(EndCertificateStatus)} value: {result.Status}");
         }
+
+        private async Task<LoadCertificatesResult> LoadCertificates(EndCertificateValidation validation)
```
[the remark this answers is not in the document]
Comment (loic-sharma) [reply src/Validation.PackageSigning.ValidateCertificate/CertificateValidationMessageHandler.cs]: Fixed

File: src/Validation.PackageSigning.ValidateCertificate/CertificateValidationMessageHandler.cs [hunk 5]
```diff
@@ -155,5 +180,31 @@ private bool HasValidationCompleted(EndCertificateValidation validation, Certifi
...
+        private class LoadCertificatesResult
+        {
+            public LoadCertificatesResult(X509Certificate2 endCertificate, X509Certificate2[] parentCertificates)
+            {
+                EndCertificate = endCertificate;
+                ParentCertificates = parentCertificates;
+            }
+
+            public X509Certificate2 EndCertificate { get; }
+            public X509Certificate2[] ParentCertificates { get; }
```
[the remark this answers is not in the document]
Comment (loic-sharma) [reply src/Validation.PackageSigning.ValidateCertificate/CertificateValidationMessageHandler.cs]: Good idea. Fixed

File: src/Validation.PackageSigning.ValidateCertificate/CertificateValidationService.cs (hunk 1 above)
Comment (loic-sharma) [reply src/Validation.PackageSigning.ValidateCertificate/CertificateValidationService.cs]: Very good catch, thanks! Fixed :)

File: src/Validation.PackageSigning.ValidateCertificate/CertificateValidationMessageHandler.cs [hunk 6]
```diff
@@ -157,5 +180,40 @@ private bool HasValidationCompleted(EndCertificateValidation validation, Certifi
...
+            thumbprints.AddRange(validation.EndCertificate.CertificateChainLinks.Select(l => l.ParentCertificate.Thumbprint));
+
+            var certificates = await Task.WhenAll(thumbprints.Select(t => _certificateStore.LoadAsync(t, CancellationToken.None)));
+
+            return new LoadCertificatesResult(certificates.First(), certificates.Skip(1).ToArray());
+        }
+
+        private class LoadCertificatesResult : IDisposable
+        {
+            public LoadCertificatesResult(X509Certificate2 endCertificate, X509Certificate2[] parentCertificates)
```
Comment (joelverhagen) [inline src/Validation.PackageSigning.ValidateCertificate/CertificateValidationMessageHandler.cs]: `IReadOnlyList`

File: src/Validation.PackageSigning.ValidateCertificate/SignatureDeciderFactory.cs (hunk 3 above)
Comment (loic-sharma) [reply src/Validation.PackageSigning.ValidateCertificate/SignatureDeciderFactory.cs]: Fixed

File: src/Validation.PackageSigning.ValidateCertificate/CertificateValidationMessageHandler.cs (hunk 6 above)
Comment (loic-sharma) [reply src/Validation.PackageSigning.ValidateCertificate/CertificateValidationMessageHandler.cs]: Fixed. I went ahead and updated several APIs that accepted `X509Certificate2[]` as well.

**Pull request merged**